### ISSUE TITLE
Config from `mix run --config` invocations doesn't get deeply merged

### ISSUE THREAD

**Issue by pdobacz** [elixir-lang/elixir#8811]
### Environment

```
$ elixir --version
Erlang/OTP 21 [erts-10.2.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Elixir 1.8.0 (compiled with Erlang/OTP 20)
```

* Operating system: Ubuntu 18.04

### Current behavior

When I put, e.g.:

```
config :my_app, MyApp.Endpoint,
  http: [port: 7435]
```

into a `some_config.exs` file and then `mix run --config some_config.exs`, I'll get:

```
IO.inspect Application.get_env(:my_app, MyApp.Endpoint)

# [http: [port: 7435]]
```

despite having entries in the "default" config files loaded in a rather standard manner (i.e. `config/config.exs` doing `import_config config/dev.exs` etc.).

### Expected behavior

I'd expect the config from `some_config.exs` be deeply merged as are the `config/config.exs` and `config/dev.exs` loaded by default.

That is, if without the override in `--config some_config.exs` I have:

```
IO.inspect Application.get_env(:my_app, MyApp.Endpoint)

# [
#   secret_key_base: ...
# <snip>,
#   http: [port: 7434],
# <snip>
# ]
```

**with the override `--config some_config.exs`**, I'd expect to have 

```
# [
#   secret_key_base: ...
# <snip>,
#   http: [port: 7435],
# <snip>
# ]
```

which would effectively be `config/config.exs` with `config/dev.exs` merged in with `some_config.exs` merged in, deeply

**Comment by josevalim** [comment]
Those happen at two very distinct steps and therefore we cannot properly deep merge them since the original config has been long gone. I think we have two options here:

  1. Deprecate the `--config` flag and push developers to handle those conflicts in different ways (for example, by adding extra environments)

  2. Improve the documentation to say it does not deep merge

I am more inclined to go with 1 because I beleive most would expect deep merge to work and documenting it is not enough.

**Comment by pdobacz** [comment]
1 does sound like more along the pattern here.

We're using the `--config` files to sneak bulky runtime configuration into `Application`'s env, which is probably what is against the pattern. However, despite being non-standard, it worked well, and then it's easy to run into this "no deep merge" issue.

Actually, what would be the correct and convenient way of populating `Application`'s env based on runtime information contained in a file, if `--config` is removed? I did some quick searches and found nothing relevant which would provide an end-to-end, drop-in replacement.

**Comment by josevalim** [comment]
You can use Mix.Config.load and Mix.Config.persist to achieve a similar result or just set Application.get_env directly from a script file. So you would do something like `mix run -r sets_config.exs` or similar. But if you want merging, doing it inside `config/config.exs` is the best way to go. You can even do something like:

```elixir
if extra = System.get_env("EXTRA_CONFIG") do
  import_config "#{extra}.exs"
end
```

**Comment by pdobacz** [comment]
Thank you very much! I'll give these a try when I have the chance, for now continuing with `--config` and anticipating it might be deprecated.

Not sure what should I do with this issue - close as deep-merging for `--config` is not an option or keep open to cover taking either 1 (or 2) path?

**Comment by josevalim** [comment]
We will keep it open until it is deprecated and then we will close it. Thank you!

**Comment by josevalim** [comment]
Closed in favor of #8932.
